nomos-services/chain-service/src/lib.rs

```diff
@@ -1644,3 +1656,197 @@ where
         .map_err(|(error, _)| Box::new(error) as DynError)?;
     receiver.await.map_err(|error| Box::new(error) as DynError)
 }
+
+/// Reconstruct a Block from a Proposal by looking up transactions from mempool
+async fn reconstruct_block_from_proposal<Payload, Item, Key>(
+    proposal: Proposal,
+    mempool: OutboundRelay<MempoolMsg<HeaderId, Payload, Item, Key>>,
+) -> Result<Block<Item>, Error>
+where
+    Key: Send + Eq + Hash + From<ZkHash> + Debug,
+    Payload: Send,
+    Item: Clone + Transaction<Hash = Key>,
+{
+    let tx_map: HashMap<Key, Item> = get_mempool_contents(mempool)
```

Comment:
do we dump the full mempool here to get a few items? seems a bit inefficient. One example could be to add a custom op to review txs by hash